// @grafana/faro-react/dist/cjs/router/v6/initialize.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeReactRouterV6Instrumentation = initializeReactRouterV6Instrumentation;
exports.createReactRouterV6Options = createReactRouterV6Options;
exports.initializeReactRouterV6DataRouterInstrumentation = initializeReactRouterV6DataRouterInstrumentation;
exports.createReactRouterV6DataOptions = createReactRouterV6DataOptions;
var hoist_non_react_statics_1 = __importDefault(require("hoist-non-react-statics"));
var types_1 = require("../types");
var FaroRoutes_1 = require("./FaroRoutes");
var routerDependencies_1 = require("./routerDependencies");
function initializeReactRouterV6Instrumentation(dependencies) {
    (0, hoist_non_react_statics_1.default)(FaroRoutes_1.FaroRoutes, dependencies.Routes);
    (0, routerDependencies_1.setReactRouterV6Dependencies)(dependencies);
}
function createReactRouterV6Options(dependencies) {
    return {
        version: types_1.ReactRouterVersion.V6,
        dependencies: dependencies,
    };
}
function initializeReactRouterV6DataRouterInstrumentation(dependencies) {
    (0, routerDependencies_1.setReactRouterV6DataRouterDependencies)(dependencies);
}
function createReactRouterV6DataOptions(dependencies) {
    return {
        version: types_1.ReactRouterVersion.V6_data_router,
        dependencies: dependencies,
    };
}
