// @grafana/faro-react/dist/cjs/router/v6/routerDependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useNavigationType = exports.useLocation = exports.Routes = exports.matchRoutes = exports.createRoutesFromChildren = exports.isInitialized = void 0;
exports.setReactRouterV6Dependencies = setReactRouterV6Dependencies;
exports.setReactRouterV6SSRDependencies = setReactRouterV6SSRDependencies;
exports.setReactRouterV6DataRouterDependencies = setReactRouterV6DataRouterDependencies;
exports.isInitialized = false;
function setReactRouterV6Dependencies(newDependencies) {
    exports.isInitialized = true;
    exports.createRoutesFromChildren = newDependencies.createRoutesFromChildren;
    exports.matchRoutes = newDependencies.matchRoutes;
    exports.Routes = newDependencies.Routes;
    exports.useLocation = newDependencies.useLocation;
    exports.useNavigationType = newDependencies.useNavigationType;
}
function setReactRouterV6SSRDependencies(newDependencies) {
    exports.Routes = newDependencies.Routes;
}
function setReactRouterV6DataRouterDependencies(newDependencies) {
    exports.isInitialized = true;
    exports.matchRoutes = newDependencies.matchRoutes;
}
